File: Ahorn/entities/GalactaKnightBoss.jl
module GalactaKnightBoss

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/GalactaKnightBoss" GalactaKnightBoss(x::Integer, y::Integer, health::Integer=1700, maxHealth::Integer=1700)

const placements = Ahorn.PlacementDict(
    "galacta_knight_boss" => Ahorn.EntityPlacement(GalactaKnightBoss)
)

function Ahorn.selection(entity::GalactaKnightBoss)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::GalactaKnightBoss, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/galacta_knight_boss.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local galactaKnightBoss = {}

galactaKnightBoss.name = "MaggyHelper/GalactaKnightBoss"
galactaKnightBoss.depth = -8500
galactaKnightBoss.justification = {0.5, 1.0}

galactaKnightBoss.placements = {
    {
        name = "galacta_knight_boss",
        data = {
            health = 1700,
            maxHealth = 1700,
    }
    }
}

galactaKnightBoss.fieldInformation = {
    health = {
        fieldType = "integer",
        minimumValue = 1
    },
    maxHealth = {
        fieldType = "integer",
        minimumValue = 1
    }
}

galactaKnightBoss.fieldOrder = {
    "x", "y",
    "health",
    "maxHealth"
}

function galactaKnightBoss.sprite(room, entity)
    local textures = {
        "gui/kirby/powers/knight",
        "characters/Kglobal::Player/sitDown00"
    }

    for _, texture in ipairs(textures) do
        local ok, sprite = pcall(drawableSprite.fromTexture, texture, entity)
        if ok and sprite then
            sprite:setJustification(0.5, 1.0)
            return {sprite}
        end
    end

    return {}
end

function galactaKnightBoss.selection(room, entity)
    return utils.rectangle(entity.x - 17, entity.y - 48, 34, 48)
end

return galactaKnightBoss
